linalg/hiop.hpp

```diff
@@ -0,0 +1,192 @@
+﻿// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_HIOP
+#define MFEM_HIOP
+
+#include "linalg.hpp"
+#include "../config/config.hpp"
+#include "../general/globals.hpp"
+
+#ifdef MFEM_USE_MPI
+#include "operator.hpp"
+#endif
+
+#ifdef MFEM_USE_HIOP
+
+#include "hiopInterface.hpp"
+#include "hiopNlpFormulation.hpp"
+
+namespace mfem
+{
+
+/// Adapts the OptimizationProblem class to HiOp's interface.
+class HiopOptimizationProblem : public hiop::hiopInterfaceDenseConstraints
+{
+private:
+
+#ifdef MFEM_USE_MPI
+   MPI_Comm comm_;
+#endif
+
+   // Problem info.
+   OptimizationProblem &problem;
+
+   // Local and global number of variables and constraints.
+   long long ntdofs_loc, ntdofs_glob, m_total;
+
+   // Initial guess.
+   const Vector *x_start;
+
+   Vector constr_vals;
+   DenseMatrix constr_grads;
+   bool constr_info_is_current;
+   void UpdateConstrValsGrads(const Vector x);
+
+public:
+   HiopOptimizationProblem(OptimizationProblem &prob)
+      : x_start(NULL),
+        problem(prob),
+        ntdofs_loc(prob.input_size), ntdofs_glob(ntdofs_loc),
+        m_total(prob.GetNumConstraints()),
+        constr_vals(m_total), constr_grads(), constr_info_is_current(false)
+  { 
+#ifdef MFEM_USE_MPI
+    // Used when HiOp with MPI support is called by a serial driver.
+    comm_ = MPI_COMM_WORLD;
```

Comment:
I think `MPI_COMM_SELF` might be better default behavior.

Reply:
I prefer _WORLD since it results in the same behavior with _SELF when the driver is serial, however, hiop works in parallel when the driver is mpi and this constructor gets called for some reason (like negligence)